src/storage/src/source/mod.rs

```diff
@@ -850,125 +965,553 @@ where
+        let mut remap_trace_buffer = Vec::new();
+
+        // The global view of the source_upper, which we track by combining
+        // summaries from the raw reader operators.
+        let mut global_source_upper = HashMap::new();
+        let mut untimestamped_batches = Vec::new();
+
+        let upper_ts = resume_upper.as_option().copied().unwrap();
+        let as_of = Antichain::from_elem(upper_ts.saturating_sub(1));
+        let mut timestamper = ReclockFollower::new(as_of);
+
+        move |frontiers| {
+            batch_input.for_each(|_cap, data| {
+                data.swap(&mut batch_buffer);
+                for batch in batch_buffer.drain(..) {
+                    let batch = batch
+                        .borrow_mut()
+                        .take()
+                        .expect("batch already taken, but we should be the only consumer");
+                    for (pid, offset) in batch.source_upper.iter() {
+                        let previous_offset = global_source_upper.insert(pid.clone(), *offset);
+                        if let Some(previous_offset) = previous_offset {
+                            assert!(previous_offset <= *offset);
                         }
-                    },
+                    }
+                    untimestamped_batches.push(batch);
                 }
-                if timer.elapsed() > YIELD_INTERVAL {
-                    let _ = activator.activate();
+            });
+
+            remap_input.for_each(|cap, data| {
+                data.swap(&mut remap_trace_buffer);
+                for update in remap_trace_buffer.drain(..) {
+                    timestamper.push_trace_updates(update.into_iter())
+                }
+                cap_set.insert(cap.retain());
+            });
+
+            let remap_frontier = &frontiers[1];
+            trace!(
+                "reclock({id}) {worker_id}/{worker_count}: remap frontier: {:?}",
+                remap_frontier.frontier()
+            );
+            timestamper.push_upper_update(remap_frontier.frontier().to_owned());
+
+            // Accumulate updates to bytes_read for Prometheus metrics collection
+            let mut bytes_read = 0;
+            // Accumulate updates to offsets for system table metrics collection
+            let mut metric_updates = HashMap::new();
+
+            trace!(
+                "reclock({id}) {worker_id}/{worker_count}: \
+                untimestamped_batches.len(): {}",
+                untimestamped_batches.len(),
+            );
+
+            // WIP: Be smarter about this.
+            while let Some(untimestamped_batch) = untimestamped_batches.first_mut() {
+                let reclocked = match timestamper.reclock(&mut untimestamped_batch.messages) {
+                    Ok(reclocked) => reclocked,
+                    Err((pid, offset)) => panic!("failed to reclock {} @ {}", pid, offset),
+                };
+
+                if let Some(reclocked) = reclocked {
```

Comment:
This also looks like a desugared match statement

Reply:
:+1: